pkg/datastore/dynamodb/iterator.go

```diff
@@ -0,0 +1,46 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamodb
+
+import (
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+)
+
+// Iterator is wrapper of queried data pool
+type Iterator struct {
+	datapool []map[string]*dynamodb.AttributeValue
```

Comment:
Since dynamodb doesn't support cursor pattern (it uses `ExclusiveStartKey` and `LastEvaluatedKey` for pagination), my solution is to query throw all possible items and store them to this Iterator object, then return the pointer to this datastore object and make it act as normal Iterator (with Next() function). All pagination/ordering logic could be implemented by apply page splitting/order to this `datapool`.
Please share if you have a better solution for this 😄(I'm a bit worried about the large Iterator object that we might get from this method)

Reply:
IMO, It's enough to cache all results set at the first version. But as you feel it's not safe for the case of the large results set. So I'm thinking about whether we can `ScanPagesWithContext` which is now used inside the Find() method whenever this `Iterator.Next()` is called.

https://github.com/pipe-cd/pipe/pull/1521/files#diff-fe3bf1266a3d42a83de476d37abd775d69441d467f689bb18a6964953aa0b2d9R135

Reply:
Of course, I don't mean throwing a large number of queries with a page size of 1.

Reply:
Good idea 👍 In that case `datapool` will hold (in worst) all data of one page (with specified default PageSize). I will implement it later with a separate PR.

Reply:
That's the case. That's what I was trying to say 👍